265 interaction

sibyl229 opened the pull request:
The objective is to reuse the gene page's interaction widget on wbprocess page and interaction page.

Technical summary:
- The concept of focus gene isn't applicable on other pages. Functions that rely on the focus gene are re-written to avoid focus gene
- Rewrite queries to get the set of correct tuples (`interaction-interactorholder1-interactorholder2`) at query time to avoid subsequent filtering (that is somewhat difficult to track).
- Add interactions widget for wbprocess page and interaction page
- Support string-based "other interactors"

Commit:
refactor build-interactions to take interactions as args

Commit:
avoid relying on focus gene,which is not generalizable

Commit:
add gene->interaction-3, interaction->gene-3, gene-neighbour-5

Commit:
uniquely identify an interaction row by [interaction, gene1, gene2] tuple

Commit:
non-predicted neighbour

Commit:
take direction into account when grouping rows

Commit:
include predicted nearby interactions

Commit:
refactor code that fill interaction tuples with content

Commit:
consider feature, molecule, rearrangement as interactor

Commit:
remove commented out code

Commit:
move gene.widgets.interactions to interaction.core

Commit:
fix problem with int-rules

Commit:
add interactions widget on wbprocess page

Commit:
display other interactor

Commit:
add interactors widget on interaction page

mgrbyte commented on src/rest_api/classes/interaction/core.clj:
```diff
@@ -45,16 +46,56 @@
+     [?h :interaction.rearrangement/rearrangement ?rearrangement]
+     [?int :interaction/rearrangement ?h]]
+    [(x->interaction-3 ?x ?h ?ix)
+     (or (gene->interaction-3 ?x ?h ?ix)
```
a few questions and comments...

1.  New things I see are the introduction of "type" (e.g molecule, feature) specific interaction relations - were these being found before and are now just being literally expanded for readability?
 
2. I'm not entirely sure the naming of rule-heads (having the number-of-params as a suffix as a naming thing)

3.  _seems_ counter intuitive to have an `(or` statement  (since rules of the same name naturally *or* together; but this is a very complex area, so can understand why it's easier to break the  rules up. Related to `1.` am sure.
Perhaps naming the rules individually with defs and making `int-roles` a composite of them would help?
In that way one could still keep the same "rule head" names (would perhaps remove the need to

mgrbyte reviewed (commented):
In general, code looks good to me.

Commit:
remove suffix in rules

Commit:
fix problem caused by string-based other-interactor used in a query

sibyl229 commented:
Hi @mgrbyte I noticed a bug in my code after you reviewed it. It took me a long time to realize what was wrong. Finally fixed it and, PR is ready again.

To respond to your suggestions and questions:

1)  Yes, the molecule, feature, and rearrangement interactors have always been part of this table. It's just refactoring. "Other interactor" wasn't shown before, but Chris indicated that it should be.

2) The number of parameter in the rule name is removed :)

3) I didn't know that rule heads of the same name are `or` ed together. I kinda have those type-based rule heads in case a query is only interested in getting interactors of a particular type. But, so far in the code, we haven't encountered such a case. Let me know if you have any strong feeling about this. I can rename those if you want.

mgrbyte commented:
1 &  2 :+1: 
3.
I think what I'd do here is use the same name for all rule heads, and:
 * Separate each rule out into it's own `def`  (capability to use the rules to fetch type specific interactors   
 * Make anyother def for  all-rules, that aggregates all the individual rules (current usage).

So instead of `gene->interaction`, 'molecule->interaction` et al, rule name would just be `interaction`.
but because splitting them out into different `def`s means you could still use them interdependently.

Commit:
same rule heads regardless of interactor type

sibyl229 commented:
Thanks a lot @mgrbyte ! I updated the rule heads to be the same and removed the redundant `or` clause. When a use case for individual interactor-based rule appears, I will separate them into functions.

mgrbyte reviewed (approved):
I like this much better! :+1: 
Only change I'd make is the naming:

`x->interaction to`  `->interaction`
`interaction->x` to  `interaction->`

... but that's being picky! :wink:

sibyl229 commented:
Thanks a lot @mgrbyte ! I will fix that before pulling it in. I imagine it's an idiomatic thing in Clojure

Commit:
renaming of datomic query rules heads
